File: src/main/java/io/dddrive/enums/model/base/EnumerationBase.java

package io.dddrive.enums.model.base;

import io.dddrive.app.service.api.impl.Enumerations;
import io.dddrive.enums.model.Enumerated;
import io.dddrive.enums.model.Enumeration;

import static io.dddrive.util.Invariant.assertThis;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class EnumerationBase<E extends Enumerated> implements Enumeration<E> {

	private final String module;
	private final String id;
	private List<E> items = new ArrayList<E>();
	private Map<String, E> itemsById = new HashMap<String, E>();

	public EnumerationBase(Class<E> enumeratedClass, Enumerations enums) {
		String[] parts = this.getClass().getCanonicalName().split("\\.");
		int numOfParts = parts.length;
		// assertThis(parts.length == 7,
		// () -> "valid enumeration class name i
		// ([io].[company/project].[area].[module].model.enums.[xyEnum]): "
		// + this.getClass().getCanonicalName());
		assertThis(numOfParts >= 4,
				() -> "valid enumeration class name (i), ([company/project].[module].model.enums.[xyEnum]): "
						+ this.getClass().getCanonicalName());
		assertThis("model".equals(parts[numOfParts - 3]),
				() -> "valid enumeration class name (ii), must end with (model.enums.[xyEnum]): "
						+ this.getClass().getCanonicalName());
		assertThis("enums".equals(parts[numOfParts - 2]),
				() -> "valid enumeration class name (iii), must end with (model.enums.[xyEnum]): "
						+ this.getClass().getCanonicalName());
		assertThis(parts[numOfParts - 1].endsWith("Enum"),
				() -> "valid enumeration class name (iv), must end with (model.enums.[xyEnum]): "
						+ this.getClass().getCanonicalName());
		this.module = parts[numOfParts - 4];
		this.id = Character.toLowerCase(parts[numOfParts - 1].charAt(0)) + parts[numOfParts - 1].substring(1);
		enums.addEnumeration(enumeratedClass, this);
	}

	@Override
	public String getModule() {
		return this.module;
	}

	@Override
	public String getId() {
		return this.id;
	}

	protected void addItem(E item) {
		this.items.add(item);
		this.itemsById.put(item.getId(), item);
	}

	@Override
	public List<E> getItems() {
		return this.items;
	}

	@Override
	public E getItem(String id) {
		if (id == null) {
			return null;
		}
		E item = this.itemsById.get(id);
		assertThis(item != null, "valid item [" + id + "]");
		return item;
	}

	@Override
	public String getResourcePath() {
		return this.module + "/" + this.id.replace("Enum", "");
	}

}

File: fm-dddrive/src/main/java/dddrive/ddd/model/base/EnumerationBase.kt
package dddrive.ddd.model.base

import dddrive.ddd.model.Enumerated
import dddrive.ddd.model.Enumeration
import dddrive.ddd.model.RepositoryDirectory
import dddrive.ddd.model.RepositoryDirectorySPI

abstract class EnumerationBase<E : Enumerated>(
	enumeratedClass: Class<E>,
) : Enumeration<E> {

	override lateinit var area: String
	override lateinit var module: String
	override lateinit var id: String
	override val items: MutableList<E> = mutableListOf()
	private val itemsById: MutableMap<String, E> = mutableMapOf()

	init {
		val parts = enumeratedClass
			.getCanonicalName()
			.split("\\.".toRegex())
			.dropLastWhile { it.isEmpty() }
			.toTypedArray()
		val numOfParts = parts.size

		check(parts[0] == "dddrive" || numOfParts > 5) {
			"valid enumeration class name (i), ([company/project].[area].[module].model.enums.[xyEnum]): " +
					javaClass.getCanonicalName()
		}
		check("model" == parts[numOfParts - 3]) {
			"valid enumeration class name (ii), must end with (model.enums.[xyEnum]): " +
					javaClass.getCanonicalName()
		}
		check("enums" == parts[numOfParts - 2]) {
			"valid enumeration class name (iii), must end with (model.enums.[xyEnum]): " +
					javaClass.getCanonicalName()
		}

		area = parts[numOfParts - 5]
		module = parts[numOfParts - 4]
		id = parts[numOfParts - 1][0].lowercaseChar().toString() + parts[numOfParts - 1].substring(1) + "Enum"

		(RepositoryDirectory.Companion.instance as RepositoryDirectorySPI).addEnumeration(enumeratedClass, this)
	}

	open fun addItem(item: E) {
		require(EnumConfigBase.isInConfig()) { "in ddd configuration" }
		require(itemsById[item.id] == null) { "unique item [" + item.id + "] in enumeration [" + id + "]" }
		EnumConfigBase.addEnum(this)
		items.add(item)
		itemsById[item.id] = item
	}

	open fun assignItems() {
		require(EnumConfigBase.isInConfig()) { "in ddd configuration" }
	}

	override fun getItem(id: String): E {
		val item = itemsById[id]
		check(item != null) { "valid item [" + id + "] in enumeration [" + id + "]" }
		return item
	}

	override val resourcePath: String
		get() = "$module.${id.replace("Enum", "")}"

}
